pkg/di/ir/program.go

```diff
@@ -0,0 +1,162 @@
+// is used to drive both eBPF program generation and the interpretation of
+// data that comes out of the generated program.
+type Program struct {
+
+	// ID is the ID of the program. This is used to identify messages
+	// from this program as they are communicated over the ring buffer.
+	ID ProgramID
+
+	// Probes are the set of probes that comprise the program.
+	Probes []*Probe
+
+	// Subprograms are a list of subprograms that will be probed.
+	Subprograms []*Subprogram
+
+	// Types are the types that are used in the program.
+	Types []Type
+
+	// MaxTypeID is the maximum type ID that has been assigned.
+	MaxTypeID TypeID
+}
+
+// Subprogram represents a function or method in the program.
+//
+// TODO(XXX): Define the structure here when it comes to inlining.
+type Subprogram struct {
+
+	// Name is the name of the subprogram.
+	Name string
+
+	// OutOfLinePCRanges are the ranges of PC values that will be probed
+	// for the out-of-line-instances of the subprogram.
+	// These are sorted by start PC.
+	//
+	// What does this mean for inlined subprograms?
+	OutOfLinePCRanges []PCRange
+
+	// InlinePCRanges are the ranges of PC values that will be probed
+	// for the inlined instances of the subprogram.
+	// These are sorted by start PC.
+	//
+	// TODO(XXX): Decide whether this deserves more structure.
+	InlinePCRanges [][]PCRange
+
+	// Variables are the variables that are used in the subprogram.
+	Variables []Variable
+
+	// Lines are the lines of the subprogram.
+	//
+	// TODO(XXX): Decide whether we want to keep this.
+	Lines []SubprogramLine
+}
+
+// Line represents a line in the subprogram.
+type SubprogramLine struct {
+	PC              uint64
+	File            string
+	Line            uint32
+	Column          uint32
+	IsStatement     bool
+	IsPrologueEnd   bool
+	IsEpilogueStart bool
+}
+
+// Variable represents a variable or parameter in the subprogram.
+type Variable struct {
+	// Name is the name of the variable.
+	Name string
+	// Type is the type of the variable.
+	Type Type
+	// Locations are the locations of the variable in the subprogram.
+	Locations []Location
+	// IsParameter is true if the variable is a parameter.
+	IsParameter bool
+}
+
+// Location is the location of a parameter or variable in the subprogram.
+type Location struct {
+	// PCRange is the range of PC values that will be probed.
+	Range PCRange
+	// The locations of the pieces of the parameter or variable.
+	Location []bininspect.ParameterPiece
+}
+
+// PCRange is the range of PC values that will be probed.
+type PCRange struct {
+	// Start is the start PC value of the range.
+	Start uint64
+	// End is the end PC value of the range.
+	End uint64
+}
+
+// Probe represents a probe from the config as it applies to the program.
+type Probe struct {
+	// The config UUID ID of the probe.
+	ID string
+
+	// The kind of the probe.
+	Type ProbeKind
+
+	// The version of the probe.
+	Version int
+
+	// Tags that are passed through the probe.
+	Tags []string
```

Comment:
What's an example tag?

Reply:
You tell me. It's in the API [here](https://github.com/DataDog/dd-go/blob/9bbe4002a415fd613c13ac8c986dc04c86211b22/remote-config/pkg/products/livedebugging/domain.go#L81). I have to imagine it's a way that the user or backend can ensure that the output log or metric or span carries extra tags.

Reply:
Hm ok, Tags as far as I understand are used to specify instances of services to instrument. Let's sync with someone from another runtime to ensure we're using it correctly.

Reply:
My sense is that it's these queries in Enablement that choose the services: https://github.com/DataDog/dd-go/blob/beeaadef389190e4e6bbabf82059aa0c91eba770/remote-config/apps/rc-api/products/livedebugging/jsonapiconf/domain.go#L63-L73

I think these are tags that should come out the other side. Here's an example I found in the real config:

```
  "tags": [
    "ui_session_id:<redacted>",
    "session_name:<redacted>'s session May 12, 2:07 pm"
  ],
  ```
  
  Note that I did the redacting.

Reply:
Got it, thanks for digging into it